python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +98,432 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
+    where :math:`u=(u_0,...,u_N), v=(v_0,...,v_N)` using PETSc SNES as the non-linear solver.
+
+    Note:
+        The user is responsible for the destruction of the returned objects.
+
+    Args:
+        F: List of PDE residuals :math:`F_i`.
+        u: List of unknowns of the block system.
+        bcs: List of Dirichlet boundary conditions.
+        J: Rectangular array of bi-linear forms representing the
+            Jacobian :math:`J_ij = dF_i/du_j`.
+        P: Rectangular array of bi-linear forms representing the preconditioner.
+        assembly_type: Determines what kind of matrix the problem is assembled into.
+            If `N=0`, i.e one is solving for a single form, the assembly type is `default`.
+            For blocked, the assembly type is `block` or `nest`.
+        form_compiler_options: Options used in FFCx
+            compilation of this form. Run ``ffcx --help`` at the
+            command line to see all available options.
+        jit_options: Options used in CFFI JIT compilation of C
+            code generated by FFCx. See ``python/dolfinx/jit.py``
+            for all available options. Takes priority over all other
+            option values.
+    Returns:
+        A PETSc SNES solver instance and a vector to store solutions in.
+    """
+    # Compile all forms
+    bcs = [] if bcs is None else bcs
+    form_compiler_options = {} if form_compiler_options is None else form_compiler_options
+    jit_options = {} if jit_options is None else jit_options
+
+    # Compile forms
+    residual = _create_form(F, form_compiler_options=form_compiler_options, jit_options=jit_options)
+    if J is None:
+        J = fem.forms.compute_jacobian(F, u)
+    jacobian = _create_form(J, form_compiler_options=form_compiler_options, jit_options=jit_options)
+    preconditioner = None
+    if P is not None:
+        preconditioner = _create_form(
+            P, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+
+    A, b, P, x = create_snes_matrices_and_vectors(jacobian, residual, preconditioner, assembly_type)
+    snes = PETSc.SNES().create(comm=A.comm)  # type: ignore
+
+    # Set function and Jacobian
+    if assembly_type == fem.petsc.AssemblyType.standard:
+        snes.setFunction(partial(F_standard, u, residual, jacobian, bcs), b)  # type: ignore
+        snes.setJacobian(partial(J_standard, u, jacobian, preconditioner, bcs), A, P)  # type: ignore
+    elif assembly_type == fem.petsc.AssemblyType.block:
+        snes.setFunction(partial(F_block, u, residual, jacobian, bcs), b)
+        snes.setJacobian(partial(J_block, u, jacobian, preconditioner, bcs), A, P)
+    elif assembly_type == fem.petsc.AssemblyType.nest:
+        snes.setFunction(partial(F_nest, u, residual, jacobian, bcs), b)
+        snes.setJacobian(partial(J_nest, u, jacobian, preconditioner, bcs), A, P)
+    else:
+        raise ValueError(f"Unsupported SNES type {assembly_type}")
+    return snes, x
```

Comment:
The thinking behind returning `x` is that users must not under any circumstances use the data behind `u` to call SNES `solve`. So in a sense, this is signposting that may not be necessary for users of the lower-level interface.